### PR TITLE
[Refactor] improve performance by avoiding unnecessary component detect

### PR DESCRIPTION
**Pull request by golopot**
Rules using `Component.detect()` are generally slow. Some rules uses `Component.detect(context, components, utils)` solely for the `utils` it provided. These `Component.detect`s can be avoided by extracting some `utils` functions into a module.

Benchmark with config `plugin:react/all`:
```diff
$ time npx eslint .
- 16.429
+ 13.692
```

### CODE REVIEW

File: lib/rules/jsx-no-bind.js [hunk 1]
```diff
@@ -63,7 +62,7 @@ module.exports = {
     }],
   },
 
-  create: Components.detect((context) => {
+  create(context) {
```
Comment (ljharb) [inline lib/rules/jsx-no-bind.js:65]: this doesn't seem right - we don't want this rule running on jsx that's outside of a detected component.

**Review by ljharb — CHANGES_REQUESTED** [review]
Before I review the whole thing - this seems like a very scary change to me. I'd much rather make component detection faster, and have *every* rule unconditionally be wrapped with component detection logic, than risk false positive warnings just to eke out a bit more performance from an eslint run (something that even without jest-eslint-runner is pretty fast already)

**Comment by golopot** [comment]
This change does not cause more reports or less reports. This change is a safe refactoring that does not change ouput. The `Component.detect` thing has no effect if you are not using `components` or `utils` it provides. Using `Component.detect` without using `components` or `utils` is analogous to spending lots of time computing a value and not using it.

**Comment by ljharb** [comment]
My expectation would be that using Components.detect would restrict the AST node visitors to only applying within components. If that's not how it works, then I clearly misunderstand the abstraction.

**Comment by golopot** [comment]
> My expectation would be that using Components.detect would restrict the AST node visitors to only applying within components. If that's not how it works, then I clearly misunderstand the abstraction.

Yes some rules uses `Component.detect((context) => { return { ImportDeclaration(node) {}}) }`, event though `ImportDeclaration` cannot possibly be in a component.

**Comment by ljharb** [comment]
Fair enough; I'll rereview.

File: lib/types.d.ts [hunk 2]
```diff
@@ -13,9 +13,7 @@ declare global {
   type JSXFragment = ASTNode;
   type JSXSpreadAttribute = ASTNode;
 
-  interface Context extends eslint.SourceCode {
-    getFirstTokens(node: estree.Node | ASTNode, options?: eslint.SourceCode.CursorWithCountOptions): eslint.AST.Token[];
-  }
+  type Context = eslint.Rule.RuleContext
```
Comment (ljharb) [inline lib/types.d.ts:16]: why this change?

File: lib/util/Components.js [hunk 3]
```diff
@@ -375,30 +276,6 @@ function componentRule(rule, context) {
       return isCreateElement(node, context);
```
[the remark this answers is not in the document]
Comment (golopot) [reply lib/util/Components.js:276]: Yes

File: lib/types.d.ts (hunk 2 above)
Comment (golopot) [reply lib/types.d.ts:16]: That Context type definition is a duct tape definition. It lacks many methods.

Comment (ljharb) [reply lib/types.d.ts:16]: ah, so this one includes getFirstTokens, awesome

**Review by ljharb — APPROVED** [review]
